### PR TITLE
Do not hash by message -- also, refactor event factory

### PR DESCRIPTION
**Pull request by silesky**
Finally, `EventFactory` is now shared between node and browser.

Features / Changes:
- Update messageId algorithm to be consistent with node (analytics-next-<epochdate>-uuid)
- Browser will throw an error in the EventFactory (not just in a plugin) if the event is invalid
- *alert* Validation now checks that one of userId/anonId/groupId/previousId is available and throws an error (to have consistency with the node validation). I assume that Browser will always have one of these values defined at this point (The user instance, etc) -- the only edge case would be if an instance had added identity resolution later.

### CODE REVIEW

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 1]
```diff
@@ -59,6 +58,34 @@ describe('Event Factory', () => {
       const group = factory.group('coolKidsId', { coolkids: true })
       expect(group.groupId).toEqual('coolKidsId')
     })
+
+    it('allows userId / anonymousId to be specified as an option', () => {
```
Comment (danieljackins) [inline packages/browser/src/core/events/__tests__/index.test.ts:62]: copypasta?

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 2]
```diff
@@ -59,6 +58,34 @@ describe('Event Factory', () => {
...
+    })
+
+    it('allows userId / anonymousId to be overridden', function () {
+      const group = factory.group('my_group_id', undefined, {
+        userId: 'bar',
+        anonymousId: 'foo',
+      })
+      expect(group.userId).toBe('bar')
+      expect(group.anonymousId).toBe('foo')
+    })
```
Comment (danieljackins) [inline packages/browser/src/core/events/__tests__/index.test.ts:78]: copypasta?

File: packages/core/src/events/index.ts [hunk 3]
```diff
@@ -11,25 +11,54 @@ import {
...
+  /**
+   * additional validation to run on each event
+   */
+  validateEvent?: ValidateEventFn
 }
 
 /**
  * This is currently only used by node.js, but the original idea was to have something that could be shared between browser and node.
  * Unfortunately, there are some differences in the way the two environments handle events, so this is not currently shared.
  */
```
Comment (danieljackins) [inline packages/core/src/events/index.ts:50]: Is this comment still relevant?

File: packages/browser/src/core/events/__tests__/index.test.ts (hunk 2 above)
Comment (silesky) [reply packages/browser/src/core/events/__tests__/index.test.ts:78]: Hah, good catch. I think this in master =S

File: packages/core/src/events/index.ts (hunk 3 above)
Comment (silesky) [reply packages/core/src/events/index.ts:50]: Good catch again!

**Review by chrisradek — APPROVED** [review]
